pkg/kube/controllers/statefulset/statefulset_rollout_reconciler.go

```diff
@@ -0,0 +1,225 @@
+package statefulset
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"code.cloudfoundry.org/quarks-utils/pkg/pointers"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"code.cloudfoundry.org/cf-operator/pkg/kube/apis"
+
+	"code.cloudfoundry.org/quarks-utils/pkg/config"
+	"code.cloudfoundry.org/quarks-utils/pkg/ctxlog"
+	"k8s.io/api/apps/v1beta2"
+	"k8s.io/apimachinery/pkg/runtime"
+	crc "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	rolloutStatePending       = "Pending"
+	rolloutStateCanary        = "Canary"
+	rolloutStateRollout       = "Rollout"
+	rolloutStateDone          = "Done"
+	rolloutStateFailed        = "Failed"
+	rolloutStateCanaryUpscale = "CanaryUpscale"
+)
+
+var (
+	// AnnotationCanaryRolloutEnabled if set to "true" canary behaviour is desired
+	AnnotationCanaryRolloutEnabled = fmt.Sprintf("%s/canary-rollout-enabled", apis.GroupName)
+	// AnnotationCanaryRollout is the state of the canary rollout of the stateful set
+	AnnotationCanaryRollout = fmt.Sprintf("%s/canary-rollout", apis.GroupName)
+	// AnnotationCanaryWatchTime is the max time for the canary update
+	AnnotationCanaryWatchTime = fmt.Sprintf("%s/canary-watch-time-ms", apis.GroupName)
+	// AnnotationUpdateWatchTime is the max time for the complete update
+	AnnotationUpdateWatchTime = fmt.Sprintf("%s/update-watch-time-ms", apis.GroupName)
+	// AnnotationUpdateStartTime is the timestamp when the update started
+	AnnotationUpdateStartTime = fmt.Sprintf("%s/update-start-time", apis.GroupName)
+)
+
+// NewStatefulSetRolloutReconciler returns a new reconcile.Reconciler
+func NewStatefulSetRolloutReconciler(ctx context.Context, config *config.Config, mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileStatefulSetRollout{
+		ctx:    ctx,
+		config: config,
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+}
+
+// ReconcileStatefulSetRollout reconciles an ExtendedStatefulSet object when references changes
+type ReconcileStatefulSetRollout struct {
+	ctx    context.Context
+	client crc.Client
+	scheme *runtime.Scheme
+	config *config.Config
+}
+
+// Reconcile cleans up old versions and volumeManagement statefulSet of the ExtendedStatefulSet
+func (r *ReconcileStatefulSetRollout) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+
```

Comment:
(nit: newline)

This reconciler doesn't use the 'meltdown' package to protect against too many reconciles. (https://github.com/cloudfoundry-incubator/cf-operator/blob/master/pkg/kube/controllers/boshdeployment/deployment_reconciler.go#L89-L92)

There are two ways to use meltdown, with annotations or with a status field.

I actually wonder how this worked without it, maybe our test cases are not complicated enough to trigger enough reconcile events?

The main problem I see with the state machine approach is that it relies on an annotation, while the actual cluster state might be different, since the annotation has not been updated yet.
(see "level based api" in https://book-v1.book.kubebuilder.io/basics/what_is_a_controller.html)

Should work though, if the reconciler doesn't take too long?

Reply:
Done.